### PR TITLE
Fix List offsets bug in mixed type list column in nested JSON reader

### LINKED ISSUE #12418: [BUG] `cudf::experimental::read_json` wrongly parse data when the input is invalid
From the current json test:
```
TEST_F(JsonReaderTest, JsonNestedDtypeSchema)
```

The test specifies an input:
```
std::string json_string = R"( [{"a":[123, {"0": 123}], "b":1.0}, {"b":1.1}, {"b":2.1}])";
```
Notice that the input for list `a` (lists of structs) is invalid: the first element is `123`, which is not a struct. As such, the test expects the result of parsing `a` to be `[null, 123.0]` which is reasonable.

However, if at the end of the test, we print out the result by:
```
cudf::test::print(result.tbl->get_column(0));
```
then we will get this:
```
List<Struct<float,>>:
Length : 3
Offsets : 0, 1, 1, 1
Null count: 2
001
   Struct<float,>:
   Length : 2:
   Null count: 1
   10
      10
      NULL, 123
```

The output column seems to have corrupted data. In particular, it is a lists column having only one non-null row (the first row), which in turn has only one element (offsets `0, 1]`). However, the child column has 2 elements.

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
Fixes the bug in list offsets in mixed types - string/value type with struct type.

In nested JSON reader, following json string
`[{"a":[123, {"0": 123}], "b":1.0}, {"b":1.1}, {"b":2.1}]`
should produce first column as

```
List<Struct<float,>>:
Length : 3
Offsets : 0, 2, 2, 2
Null count: 2
001
   Struct<float,>:
   Length : 2:
   Null count: 1
   10
      10
      NULL, 123
```

Depends on  #12330
closes #12418
## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by ttnghia** [comment]
Can you also update the C++ unit test in https://github.com/rapidsai/cudf/issues/12418 too? I realize that python tests are updated in this PR but C++ tests are also need to be updated.

File: cpp/src/io/json/json_column.cu [hunk 1]
```diff
@@ -195,6 +197,45 @@ reduce_to_column_tree(tree_meta_t& tree,
...
     });
 
+  // Mixed types in List children go to different columns,
+  // so all immediate children of list column should have same max_row_offsets.
+  //   create list's children max_row_offsets array. (initialize to zero)
+  //   atomicMax on  children max_row_offsets array.
+  //   gather the max_row_offsets from children row offset array.
+  {
+    rmm::device_uvector<NodeIndexT> list_parents_children_max_row_offsets(num_columns, stream);
+    thrust::fill(rmm::exec_policy(stream),
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:207]:
```suggestion
    thrust::uninitialize_fill(rmm::exec_policy(stream),
```

File: cpp/src/io/json/json_column.cu [hunk 2]
```diff
@@ -686,7 +744,11 @@ std::pair<std::unique_ptr<column>, std::vector<column_name_info>> device_json_co
       auto [child_column, names] =
         json_col.child_columns.empty()
           ? std::pair<std::unique_ptr<column>,
-                      std::vector<column_name_info>>{std::make_unique<column>(), {}}
+                      std::vector<column_name_info>>{std::make_unique<column>(
+                                                       data_type{type_id::INT8},
+                                                       0,
+                                                       rmm::device_buffer{0, stream, mr}),
```
Comment (ttnghia) [inline cpp/src/io/json/json_column.cu:750]: Oh why? The lists column has empty child, right?

File: cpp/src/io/json/json_column.cu (hunk 1 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:207]: Difference between `thrust::fill` and [thrust::uninitialized_fill](https://thrust.github.io/doc/group__filling_ga7afcf37f1bd2799befe7e93aae7b43a1.html) is that `thrust::fill` uses assignment operator, and `thrust::uninitialized_fill` uses copy constructor. 
For primitive types, it does not make any difference.

File: cpp/src/io/json/json_column.cu (hunk 2 above)
Comment (ttnghia) [reply cpp/src/io/json/json_column.cu:750]: I don't see this change making any difference, except that the new child type is now `INT8` instead of `EMPTY`. This may avoid exception thrown by `type_dispatcher` in some code paths. But IMO I think keeping the return type of `EMPTY` column is more appropriate.

Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:750]: Type dispatcher fails when the child column is empty. That's why it's defaulted to INT8.

**Comment by karthikeyann** [comment]
> Can you also update the C++ unit test in #12418 too? I realize that python tests are updated in this PR but C++ tests are also need to be updated.
> 
> Note that I disabled that test before (in #12370). So we need to both update its content and enable it.

Enabled and updated.

**Review by ttnghia — APPROVED** [review]

File: python/cudf/cudf/tests/test_json.py [hunk 3]
```diff
@@ -1006,3 +1006,95 @@ def test_json_round_trip_gzip():
         fo.seek(loc)
         written_df = cudf.read_json(fo, orient="records", lines=True)
     assert_eq(written_df, df)
+
+
+@pytest.mark.parametrize(
+    "jsonl_string",
+    [
+        # simple list with mixed types
+        """{"a":[123, {"0": 123}], "b":1.0}\n {"b":1.1}\n {"b":2.1}""",
```
Comment (bdice) [inline python/cudf/cudf/tests/test_json.py:1015]: Why do these have newlines _and_ spaces separating the lines? Should the spaces be removed?
```suggestion
        """{"a":[123, {"0": 123}], "b":1.0}\n{"b":1.1}\n{"b":2.1}""",
```

File: python/cudf/cudf/tests/test_json.py [hunk 4]
```diff
@@ -1006,3 +1006,95 @@ def test_json_round_trip_gzip():
...
+                    else x
+                )
+        return df
+
+    # both json lines and json string tested.
+    json_string = "[" + jsonl_string.replace("\n", ",") + "]"
+    pdf = pd.read_json(jsonl_string, orient="records", lines=True)
+    pdf2 = pd.read_json(json_string, orient="records", lines=False)
+    assert_eq(pdf, pdf2)
+    pdf = _replace_with_nulls(pdf, [123, "123", 12.3])
```
Comment (vuule) [inline python/cudf/cudf/tests/test_json.py:1079]: can you please document why this modification is needed?

**Review by bdice — APPROVED** [review]
Approving Python test.

File: python/cudf/cudf/tests/test_json.py (hunk 3 above)
Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:1015]: Spaces does not matter.

File: cpp/src/io/json/json_column.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/io/json/json_column.cu:751]: The `EMPTY` type is likely to be removed: https://github.com/rapidsai/cudf/issues/12477

Comment (ttnghia) [reply cpp/src/io/json/json_column.cu:751]: > Type dispatcher fails when the child column is empty. That's why it's defaulted to INT8.

Can you also add a comment quoting this please?

**Comment by vuule** [comment]
@karthikeyann could you summarize the fix in the PR description? I'm having a hard time following the logic of the code changes, so it's probably a good idea to leave some additional info for future development.

File: python/cudf/cudf/tests/test_json.py [hunk 5]
```diff
@@ -1006,3 +1006,95 @@ def test_json_round_trip_gzip():
         fo.seek(loc)
         written_df = cudf.read_json(fo, orient="records", lines=True)
     assert_eq(written_df, df)
+
+
+@pytest.mark.parametrize(
+    "jsonl_string",
+    [
```
Comment (vuule) [inline python/cudf/cudf/tests/test_json.py:1013]: What happens when mixed types include both structs and lists? Example:
`{"a":[123, [{"0": 123}, {}], {"1": 321}], "b":1.0}`

File: cpp/src/io/json/json_column.cu [hunk 6]
```diff
@@ -531,6 +532,63 @@ void make_device_json_column(device_span<SymbolT const> input,
...
+     num_list_children] __device__(size_type i) {
+      auto const node_id        = node_ids[i];
+      auto const parent_node_id = parent_node_ids[node_id];
+      if (i == 0 or parent_node_ids[node_ids[i - 1]] != parent_node_id) {
+        d_columns_data[parent_col_ids[i]].child_offsets[row_offsets[parent_node_id]] =
+          row_offsets[node_id];
+      }
+      if (i == num_list_children - 1 or parent_node_ids[node_ids[i + 1]] != parent_node_id) {
+        d_columns_data[parent_col_ids[i]].child_offsets[row_offsets[parent_node_id] + 1] =
+          row_offsets[node_id] + 1;
```
Comment (vuule) [inline cpp/src/io/json/json_column.cu:588]: I think this part deserves a few comments, not sure why they were removed.

File: cpp/src/io/json/json_column.cu [hunk 7]
```diff
@@ -531,6 +532,63 @@ void make_device_json_column(device_span<SymbolT const> input,
...
 
+  // 4. scatter List offset
+  // copy_if only node's whose parent is list, (node_id, parent_col_id)
+  // stable_sort by parent_col_id of {node_id}.
+  // For all unique parent_node_id of (i==0, i-1!=i), write start offset.
+  //                                  (i==last, i+1!=i), write end offset.
+  //    unique_copy_by_key {parent_node_id} {row_offset} to
+  //    col[parent_col_id].child_offsets[row_offset[parent_node_id]]
+
+  auto& parent_col_ids = sorted_col_ids;  // reuse sorted_col_ids
```
Comment (vuule) [inline cpp/src/io/json/json_column.cu:543]: I'd prefer a copy here, any use of `sorted_col_ids` after alterations to `parent_col_ids` could lead to a bug. I'm more worried about such a bug than the cost of memcpy, but folks might disagree.
Maybe a future refactor (where each step is a function) could facilitate this reuse.

File: cpp/src/io/json/json_column.cu [hunk 8]
```diff
@@ -531,6 +532,63 @@ void make_device_json_column(device_span<SymbolT const> input,
...
+                             node_ids.begin());
+  thrust::for_each_n(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    num_list_children,
+    [node_ids        = node_ids.begin(),
+     parent_node_ids = tree.parent_node_ids.begin(),
+     parent_col_ids  = parent_col_ids.begin(),
+     row_offsets     = row_offsets.begin(),
+     d_columns_data  = d_columns_data.begin(),
```
Comment (vuule) [inline cpp/src/io/json/json_column.cu:578]:
```suggestion
    [node_ids        = node_ids.cbegin(),
     parent_node_ids = tree.parent_node_ids.cbegin(),
     parent_col_ids  = parent_col_ids.cbegin(),
     row_offsets     = row_offsets.cbegin(),
     d_columns_data  = d_columns_data.begin(),
```

File: python/cudf/cudf/tests/test_json.py [hunk 9]
```diff
@@ -1006,3 +1006,95 @@ def test_json_round_trip_gzip():
...
+        """{"a":[{"0": 123}, "123"], "b":1.0}\n {"b":1.1}\n {"b":2.1}""",
+        """{"a":["123", {"0": 123}, "123"], "b":1.0}\n {"b":1.1}""",
+        """{"a":[123]}\n {"a":[{"0": 123}], "b":1.0}\n {"b":1.1}""",
+        """{"a":[{"0": 123}]}\n {"a":[123], "b":1.0}\n {"b":1.1}""",
+        """{"a":[{"0": 123}]}\n {"a": []}\n {"a":[123], "b":1.0}\n{"b":1.1}""",
+        """{"b":1.0, "a":[{"0": 123}]}\n {"a":[123]}\n {"b":1.1}\n{"a": []}""",
+        """{"a": []}\n {"a":[{"0": 123}]}\n {"a":[123], "b":1.0}\n{"b":1.1}""",
+        """{"a": []}\n {"a":[123], "b":1.0}\n {"a":[{"0": 123}]}\n{"b":1.1}""",
+        # nested list with mixed types
+        """{"a":[123, [{"0": 123}, {}]], "b":1.0}
```
Comment (vuule) [inline python/cudf/cudf/tests/test_json.py:1029]: this might be a separate case:
`{"a":[123, {}], "b":1.0}`

**Review by vuule — COMMENTED** [review]
Minor comments, still wrapping my head around the logic of the changes (hoping requested comments/description can help).
Tried to come up with more test cases anyway :)

File: cpp/src/io/json/json_column.cu (hunk 8 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:578]: `cbegin()` does not exist for device_span.

Comment (vuule) [reply cpp/src/io/json/json_column.cu:578]: Have to admit, I didn't check types of all inputs here, suggestion is to use cbegin where applicable :)
Sorry for the extra churn this causes.

File: python/cudf/cudf/tests/test_json.py (hunk 5 above)
Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:1013]: It's an error. It will throw an exception.

File: cpp/src/io/json/json_column.cu (hunk 2 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:751]: @ttnghia as discussed offline, gather throws error when children is type_id::EMPTY, during `purge_nonempty_nulls`.
Also, For expecting future gather operations on the resulting list column, we are choosing to return INT8 of size 0 here.


Also, `purge_nonempty_nulls` could be improved to skip the gather if it does not have non-empty nulls.

File: cpp/src/io/json/json_column.cu (hunk 6 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:588]: added again.

File: python/cudf/cudf/tests/test_json.py [hunk 10]
```diff
@@ -1006,3 +1006,95 @@ def test_json_round_trip_gzip():
...
+        """{"a":[123, {"0": 123}, null], "b":1.0}\n {"b":1.1}\n {"b":2.1}""",
+        """{"a":["123", {"0": 123}], "b":1.0}\n {"b":1.1}\n {"b":2.1}""",
+        """{"a":[{"0": 123}, "123"], "b":1.0}\n {"b":1.1}\n {"b":2.1}""",
+        """{"a":["123", {"0": 123}, "123"], "b":1.0}\n {"b":1.1}""",
+        """{"a":[123]}\n {"a":[{"0": 123}], "b":1.0}\n {"b":1.1}""",
+        """{"a":[{"0": 123}]}\n {"a":[123], "b":1.0}\n {"b":1.1}""",
+        """{"a":[{"0": 123}]}\n {"a": []}\n {"a":[123], "b":1.0}\n{"b":1.1}""",
+        """{"b":1.0, "a":[{"0": 123}]}\n {"a":[123]}\n {"b":1.1}\n{"a": []}""",
+        """{"a": []}\n {"a":[{"0": 123}]}\n {"a":[123], "b":1.0}\n{"b":1.1}""",
+        """{"a": []}\n {"a":[123], "b":1.0}\n {"a":[{"0": 123}]}\n{"b":1.1}""",
```
[the remark this answers is not in the document]
Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:1027]: added

File: python/cudf/cudf/tests/test_json.py (hunk 9 above)
Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:1029]: added

File: python/cudf/cudf/tests/test_json.py (hunk 4 above)
Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:1079]: described at beginning of the test. I added another comment here too.

File: cpp/src/io/json/json_column.cu (hunk 7 above)
Comment (karthikeyann) [reply cpp/src/io/json/json_column.cu:543]: This is done to reduce memory usage.
Right now, for 1 GB JSON string, it takes 4.93 GB during`reduce_to_column_tree`. It's less compared to peak memory 8.9 GB.

Comment (vuule) [reply cpp/src/io/json/json_column.cu:543]: I'm not quite sold on it, it's only an integer per column :)
